mlflow/langchain/_langchain_autolog.py

```diff
@@ -33,6 +37,144 @@
+        with disable_autologging():
+            return original(self, *args, **kwargs)
+
+    # Traces does not require an MLflow run, only the other optional artifacts require it.
+    if config.should_log_optional_artifacts():
+        with _setup_autolog_run(config, self) as run_id:
+            result = _invoke(self, *args, **kwargs)
+        _log_optional_artifacts(config, run_id, result, self, func_name, *args, **kwargs)
+    else:
+        result = _invoke(self, *args, **kwargs)
+
+    mlflow_tracer.flush_tracker()
+    return result
+
+
+@contextlib.contextmanager
+def _setup_autolog_run(config, model):
+    """Set up autologging run and return the run ID.
+
+    This function only creates a run when there is no active run and the model does not have
+    a run ID attribute propagated from the previous call. Iff it creates a new run, MLflow should
+    terminate the run at the end of the inference.
+
+    Args:
+        config: AutoLoggingConfig: The autologging configuration.
+        model: Any: The LangChain model instance that runs the inference.
+
+    Returns: yields the run IDs
+    """
+    if propagated_run_id := getattr(model, "run_id", None):
+        # When model has "run_id" attribute, it means the model is already invoked once with autolog
+        # enabled and the run_id is propagated from the previous call, so we don't create a new run.
+        run_id = propagated_run_id
+        # The run should be already terminated at the end of the previous call.
+        should_terminate_run = False
+
+    elif active_run := mlflow.active_run():
+        run_id = active_run.info.run_id
+        tags = _resolve_tags(config.extra_tags, active_run)
+        mlflow.MlflowClient().log_batch(run_id, tags=[RunTag(k, str(v)) for k, v in tags.items()])
+        should_terminate_run = False
+    else:
+        from mlflow.tracking.fluent import _get_experiment_id
+
+        run = mlflow.MlflowClient().create_run(
+            experiment_id=_get_experiment_id(),
+            run_name="langchain-" + name_utils._generate_random_name(max_length=7),
+            tags=_resolve_tags(config.extra_tags),
+        )
+        run_id = run.info.run_id
+        should_terminate_run = True
+
+    run_status = None
+    try:
+        yield run_id
+    except Exception:
+        run_status = RunStatus.to_string(RunStatus.FAILED)
+        raise
+    finally:
+        if should_terminate_run:
+            mlflow.MlflowClient().set_terminated(run_id, status=run_status)
+
+
+def _resolve_tags(extra_tags, active_run=None):
+    resolved_tags = context_registry.resolve_tags(extra_tags)
+    tags = _resolve_extra_tags(mlflow.langchain.FLAVOR_NAME, resolved_tags)
+    if active_run:
+        # Some context tags like mlflow.runName are immutable once logged, but they might be already
+        # set when the run is created, then we should avoid updating them.
+        excluded_tags = set(active_run.data.tags.keys()) & IMMUTABLE_TAGS
```

Comment:
Should any tag that starts with the protected 'namespace' `mlflow.` be considered immutable? We have similar precedent with reserving tags like this elsewhere in MLflow and I'm just curious what your thoughts are about expanding the denylist here to not overwrite tags that may be added in the future without having the context of this denylist that is reduced in scope

Reply:
> Should any tag that starts with the protected 'namespace' mlflow. be considered immutable?

Strictly to say, only a subset of reserved tags are immutable in the Databricks backend (defined as `userImmutableRunTags` in universe). That said, excluding `mlflow.` should not harm practically, as it seems none of the reserved tags need to be altered here in autologging. It sounds better than defining another allowlist in OSS side.